File: files/_contagion.py

def path(nde,toVisit,arrival,d):
    node=nde
    while node>=0:
        toVisit.remove(node)
        accMin,acc=-1,-1
        for e in toVisit:
            dist=(d[e][0]-d[node][0])*(d[e][0]-d[node][0])+(d[e][1]-d[node][1])*(d[e][1]-d[node][1])
            if arrival[e]==-1 or arrival[node]+dist<arrival[e]:
                arrival[e]=arrival[node]+dist
            if accMin==-1 or arrival[e]<accMin:
                accMin=arrival[e]
                acc=e
        node=acc
def good(place,ll,thing):
    if place>len(ll):return False
    else: return thing>=ll[place-1]
def find(x,ll):
    place=0
    jump=len(ll)
    while jump>0:
        if good(place+jump,ll,x):
            place+=jump
        jump/=2
    while good(place+1,ll,x):place+=1
    return place

def main():
    import sys
    i=lambda:sys.stdin.readline().strip()
    pipefw=lambda n,f:f(n)
    sys.setrecursionlimit(999999)

#just reading input, nothing sketchy here
    n=pipefw(i(),int)
    countries=[tuple(map(int,i().split(" "))) for _ in xrange(n)]
    start=pipefw(i(),int)-1
    q=pipefw(i(),int)
    queries=[pipefw(i(),int) for _ in xrange(q)]

#set up the djikstras
    enum=lambda e:xrange(len(e))

    arrival=[-1]*n      #time arrived for each country
    arrival[start]=0
#visited=[False]*n
    toVisit=range(n)

#do it
    path(start,toVisit,arrival,countries)

#print adj
#print arrival
    arrival.sort()
    for e in queries:
        print find(e,arrival)
#print find(input(),[(1,0),(4,0),(5,0),(6,0),(7,0),(11,0),(13,0)])
main()

File: files/_contagion.ml
open Scanf
open Printf
let (++) a b=Int64.add a b
let (--) a b=Int64.sub a b
let ( ** ) a b=Int64.mul a b
let cast=Int64.of_int;;

let n=scanf "%i\n" (fun e->e)
let cities=Array.init n (fun e->scanf "%i %i\n" (fun a b->(cast a,cast b)))
let start=scanf "%i\n" (fun e->e-1)
let q=scanf "%i" (fun e->e)
let queries=Array.init q (fun ee->scanf "\n%Ld" (fun e->e));;

let dist (a,b) (c,d)=(c--a)**(c--a)++(b--d)**(b--d);;
let vv=Array.init n (fun e-> -1L);;
vv.(start)<-0L;;
(*djikstra's algorithm*)
let rec path node toVisit=
 match toVisit with
 |[]->()
 |_->
  let ll=ref [] in
  let accMin,acc=ref (cast(-1)),ref(-1) in
  List.iter (fun place->
    let newPath=vv.(node)++(dist cities.(node) cities.(place)) in
	if vv.(place)= -1L || vv.(place)>newPath then(
	 vv.(place)<-newPath
	);
	if !accMin= -1L || vv.(place)< !accMin
	 then
	  (ll:=place:: !ll;accMin:=vv.(place);acc:=place)
	 else
	  (ll:=place:: !ll)
  ) (List.filter ((<>) node) toVisit);
  path !acc !ll;;
let rec range=function
|0->[]
|n->(n-1)::(range (n-1));;
path start (range n);;
Array.sort (fun a b->match () with () when a<b-> -1|() when a=b->0|() when a>b->1) vv;;

let rec find x ll=
 let good place thing=
  if place>Array.length ll then false else thing>=ll.(place-1)
  in
 let jump,place=ref (Array.length ll),ref 0 in
 while !jump>0 do
  if good (!place+ !jump) x then
   place:= !place+ !jump;
  jump:= !jump/2
 done;
 while good (!place+1) x do place:= !place +1 done;
 !place;;
 
Array.iter (fun e->
 find e vv |> printf "%i\n"
) queries;;
(*
Array.iter (printf "%i  ") vv;;
printf "\n";;
*)